### PR TITLE
fix: do not remove deletion markers when window time range overlaps

### PR DESCRIPTION
**Pull request by v0y4g3r**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## What's changed and what's your intention?

This PR disables deletion marker removal when some window's time range overlaps with another window, so that deletion markers will not be removed, which may cause deleted rows appear again.

## Checklist

- [x]  I have written the necessary rustdoc comments.
- [x]  I have added the necessary unit tests and integration tests.
- [x]  This PR does not require documentation updates.

### CODE REVIEW

File: src/mito2/src/compaction/twcs.rs [hunk 1]
```diff
@@ -195,24 +201,88 @@ impl Picker for TwcsPicker {
...
+    }
+
+    let mut windows = windows.into_values().collect::<Vec<_>>();
+    windows.sort_unstable_by(|l, r| l.start.cmp(&r.start).then(l.end.cmp(&r.end).reverse()));
+
+    let mut current_range: Option<(Timestamp, Timestamp)> =
+        Some((windows[0].start.unwrap(), windows[0].end.unwrap()));
+
+    for idx in 1..windows.len() {
+        match current_range {
```
Comment (evenyag) [inline src/mito2/src/compaction/twcs.rs:260]: Looks like that current_range won't be None

File: src/mito2/src/compaction/twcs.rs [hunk 2]
```diff
@@ -195,24 +201,88 @@ impl Picker for TwcsPicker {
...
 }
 
+#[derive(Default)]
+struct Window {
+    start: Option<Timestamp>,
+    end: Option<Timestamp>,
+    files: Vec<FileHandle>,
+    time_window: i64,
+    overlapping: bool,
+}
```
Comment (evenyag) [inline src/mito2/src/compaction/twcs.rs:211]: nitpick: We can define a constructor `Window::new(file: FileHandle)` and remove `default()` so files are not empty
```suggestion
struct Window {
    start: Timestamp,
    end: Timestamp,
    files: Vec<FileHandle>,
    time_window: i64,
    overlapping: bool,
}
```

File: src/mito2/src/compaction/twcs.rs [hunk 3]
```diff
@@ -195,24 +201,88 @@ impl Picker for TwcsPicker {
...
+                current_range = Some(windows[idx - 1].range());
+                continue;
+            }
+            Some(current) => {
+                let next_range = windows[idx].range();
+                if overlaps(&current, &next_range) {
+                    windows[idx - 1].overlapping = true;
+                    windows[idx].overlapping = true;
+                }
+                current_range = Some((current.0.min(next_range.0), current.1.max(next_range.1)));
```
Comment (evenyag) [inline src/mito2/src/compaction/twcs.rs:271]: I guess we should set the range to None or the next range if `overlaps()` returns false. Do unit tests cover this?

File: src/mito2/src/compaction/twcs.rs (hunk 2 above)
Comment (v0y4g3r) [reply src/mito2/src/compaction/twcs.rs]: fixed in 2b4a00cf0f

**Review by evenyag — APPROVED** [review]

**Review by MichaelScofield — APPROVED** [review]

**Pull request merged**